### PR TITLE
[thci] add Thread 1.2 Certification fixes

### PR DESCRIPTION
**Pull request by edmont**
Several fixes for Thread 1.2 Certification, including:

- getMAC
- setPollingRate
- joinCommissioned
- MGMT_ACTIVE_SET
- setCSLtout
- LinkMetricsMgmtReq
- send_udp
- setCSLsuspension
- set_max_addrs_per_child
- getMlrLogs
- mldv2_query
- mdns_query
- powerDown
- powerUp

### CODE REVIEW

File: tools/harness-thci/OpenThread.py [hunk 1]
```diff
@@ -1549,6 +1552,7 @@ def setDefaultValues(self):
         self.activetimestamp = ModuleHelper.Default_ActiveTimestamp
         # self.sedPollingRate = ModuleHelper.Default_Harness_SED_Polling_Rate
         self.__sedPollPeriod = 3 * 1000  # in milliseconds
+        self.ssedTimeout = 30
```
Comment (simonlingoogle) [inline tools/harness-thci/OpenThread.py:1555]: nit: comment the time unit?

File: tools/harness-thci/OpenThread.py [hunk 2]
```diff
@@ -2473,6 +2481,13 @@ def joinCommissioned(self, strPSKd='THREADJPAKETEST', waitTime=20):
             False: fail to start joiner
         """
         self.log("joinCommissioned on channel %s", self.getChannel())
+
+        if self.deviceRole in [
+                Thread_Device_Role.Leader,
+                Thread_Device_Role.Router,
+                Thread_Device_Role.REED,
+        ]:
+            self.__setRouterSelectionJitter(1)
```
Comment (simonlingoogle) [inline tools/harness-thci/OpenThread.py:2490]: sgtm. Even though I think it's fine to setup `routerselectionjitter` for other roles.

File: tools/harness-thci/OpenThread.py [hunk 3]
```diff
@@ -2789,7 +2815,16 @@ def MGMT_ACTIVE_SET(
 
                 cmd += policy
 
-                cmd += str(hex(policyBits))[2:]
+                flags0 = str(hex(policyBits & 0x00ff))[2:]
+                if len(flags0) < 2:
+                    flags0 = flags0.ljust(2, '0')
```
Comment (simonlingoogle) [inline tools/harness-thci/OpenThread.py:2820]:
```suggestion
                flags0 = ('%x' % (policyBits & 0x00ff)).ljust(2, '0')
```

File: tools/harness-thci/OpenThread.py [hunk 4]
```diff
@@ -2789,7 +2815,16 @@ def MGMT_ACTIVE_SET(
...
+                cmd += flags0
+
+                if self.DeviceCapability != DevCapb.V1_1:
+                    flags1 = str(hex((policyBits & 0xff00) >> 8))[2:]
+                    if len(flags1) < 2:
+                        flags1 = flags1.ljust(2, '0')
```
Comment (simonlingoogle) [inline tools/harness-thci/OpenThread.py:2826]: ditto

File: tools/harness-thci/OpenThread_BR.py [hunk 5]
```diff
@@ -465,14 +481,18 @@ def __getEthGUA(self, filterByPrefix=None):
...
-            addr = ModuleHelper.GetFullIpv6Address(addr).lower()
+            addr = str(ipaddress.IPv6Address(addr.decode()).exploded)
             globalAddrs.append(addr)
 
         if not filterByPrefix:
             return globalAddrs[0]
         else:
+            if filterByPrefix[-2:] != '::':
+                filterByPrefix = '%s::' % filterByPrefix
+            prefix = ipaddress.IPv6Address(filterByPrefix.decode())
```
Comment (simonlingoogle) [inline tools/harness-thci/OpenThread_BR.py:492]:
```suggestion
            prefix = ipaddress.IPv6Network((filterByPrefix+'/64').decode())
```
seems better to use IPv6Network

File: tools/harness-thci/OpenThread_BR.py [hunk 6]
```diff
@@ -465,14 +481,18 @@ def __getEthGUA(self, filterByPrefix=None):
...
             for fullIp in globalAddrs:
-                if fullIp.startswith(filterByPrefix):
+                address = ipaddress.IPv6Address(fullIp.decode())
+                if address.packed[:8] == prefix.packed[:8]:
```
Comment (simonlingoogle) [inline tools/harness-thci/OpenThread_BR.py:495]:
```suggestion
                if address in prefix:
```
if prefix is `ipaddress.IPv6Network`, we can simply use `in`.

**Review by simonlingoogle — APPROVED** [review]
Great work. 
LGTM 👍 
with nits

Thanks.

File: tools/harness-thci/OpenThread.py [hunk 7]
```diff
@@ -3186,7 +3222,7 @@ def setCSLperiod(self, period=500):
 
     @staticmethod
     def getForwardSeriesFlagsFromHexStr(flags):
-        hexFlags = int(flags, 16)
+        hexFlags = int(flags, 16) if isinstance(flags, str) else flags
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread.py:3225]: Probably we can rename this method since `flags` is allowed to be a hex number now.

**Review by Irving-cl — APPROVED** [review]
LGTM

File: tools/harness-thci/OpenThread.py (hunk 7 above)
Comment (edmont) [reply tools/harness-thci/OpenThread.py:3221]: Done.

File: tools/harness-thci/OpenThread_BR.py (hunk 6 above)
Comment (edmont) [reply tools/harness-thci/OpenThread_BR.py]: Done.

File: tools/harness-thci/OpenThread_BR.py (hunk 5 above)
Comment (edmont) [reply tools/harness-thci/OpenThread_BR.py]: Done.

File: tools/harness-thci/OpenThread.py (hunk 4 above)
Comment (edmont) [reply tools/harness-thci/OpenThread.py]: Done.

File: tools/harness-thci/OpenThread.py (hunk 3 above)
Comment (edmont) [reply tools/harness-thci/OpenThread.py]: Done.

File: tools/harness-thci/OpenThread.py (hunk 1 above)
Comment (edmont) [reply tools/harness-thci/OpenThread.py]: Done.

**Comment by suveshpratapa** [comment]
Hi @edmont near line 1350 of `OpenThread.py` there is the `powerDown` function.  I noticed it doesn't have the following check:

`if TESTHARNESS_VERSION == TESTHARNESS_1_2 and not self.IsBorderRouter:`

which is present in https://github.com/edmont/ot-reference-release/blob/ncs_dev/thci/OTNCS.py#L1349

Any reason we don't want to stage this change?  I just wanted to understand the reason for the version check.

**Comment by edmont** [comment]
@suveshpratapa check the discussion here: https://github.com/openthread/openthread/pull/7113#discussion_r737973839

**Review by jwhui — APPROVED** [review]

**Pull request merged**